src/algebra/lie/from_cartan_matrix.lean

```diff
@@ -0,0 +1,146 @@
+  [H_i, F_j] &= -A_{ij}F_j\\
+  ad(E_i)^{1 - A_{ij}}(E_j) &= 0 \quad\mbox{if $i \ne j$}\\
+  ad(F_i)^{1 - A_{ij}}(F_j) &= 0 \quad\mbox{if $i \ne j$}\\
+\end{align}
+$$
+
+In this file we provide the above construction. It is defined for any matrix of integers but the
+results for non-Cartan matrices should be regarded as junk.
+
+## Alternative construction
+
+This construction is sometimes performed within the free unital associative algebra
+`free_algebra R X`, rather than within the free Lie algebra `free_lie_algebra R X`, as we do here.
+However the difference is illusory since the construction stays inside the Lie subalgebra of
+`free_algebra R X` generated by `X`, and this is naturally isomorphic to `free_lie_algebra R X`
+(though the proof of this seems to require Poincaré–Birkhoff–Witt).
+
+## References
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 7--9*](bourbaki1975b) chapter VIII, §4.3
+
+* [J.P. Serre, *Complex Semisimple Lie Algebras*](serre1965) chapter VI, appendix
+
+## Main definitions
+
+  * `matrix.to_lie_algebra`
+
+## Tags
+
+lie algebra, semi-simple, cartan matrix
+-/
+
+universes u v w
+
+noncomputable theory
+
+variables (R : Type u) {B : Type v} [comm_ring R] [decidable_eq B] [fintype B]
+variables (A : matrix B B ℤ)
+
+namespace cartan_matrix
+
+variables (B)
+
+/-- The generators of the free Lie algebra from which we construct the Lie algebra of a Cartan
+matrix as a quotient. -/
+inductive generators
+| H : B → generators
+| E : B → generators
+| F : B → generators
+
+instance [inhabited B] : inhabited (generators B) := ⟨generators.H $ default B⟩
+
+variables {B}
+
+namespace relations
+
+open function
+
+local notation `H` := free_lie_algebra.of R ∘ generators.H
+local notation `E` := free_lie_algebra.of R ∘ generators.E
+local notation `F` := free_lie_algebra.of R ∘ generators.F
+local notation `ad` := lie_algebra.ad R (free_lie_algebra R (generators B))
+
+/-- The terms correpsonding to the `⁅H, H⁆`-relations. -/
+def HH : B × B → free_lie_algebra R (generators B) :=
+uncurry $ λ i j, ⁅H i, H j⁆
+
+/-- The terms correpsonding to the `⁅E, F⁆`-relations. -/
+def EF : B × B → free_lie_algebra R (generators B) :=
+uncurry $ λ i j, if i = j then ⁅E i, F i⁆ - H i else ⁅E i, F j⁆
+
+/-- The terms correpsonding to the `⁅H, E⁆`-relations. -/
+def HE : B × B → free_lie_algebra R (generators B) :=
+uncurry $ λ i j, ⁅H i, E j⁆ - (A i j) • E j
+
+/-- The terms correpsonding to the `⁅H, F⁆`-relations. -/
+def HF : B × B → free_lie_algebra R (generators B) :=
+uncurry $ λ i j, ⁅H i, F j⁆ + (A i j) • F j
+
+/-- The terms correpsonding to the `ad E`-relations.
+
+Note that we use `int.to_nat` so that we can take the power and that we do not bother
+restricting to the case `i ≠ j` since these relations are zero anyway. We also defensively
+ensure this with `ad_E_of_eq_eq_zero`. -/
+def ad_E : B × B → free_lie_algebra R (generators B) :=
+uncurry $ λ i j, (ad (E i))^(-A i j).to_nat $ ⁅E i, E j⁆
```

Comment:
This is a trick to make the extra relations when `i = j` equal to zero.

Without it, we would have an incorrect relation as shown here:
```lean
example (i : B) (h : A i i = 2) : ((ad (E i))^(1 - A i i).to_nat $ E i) = E i :=
have h' : (1 - A i i).to_nat = 0, { rw h, refl, },
by simp [h']
```

I tried to emphasise this in the doc string but I'm happy to adjust / change the implementation / reword the doc string if you have suggestions!

Reply:
Perhaps I should add that I considered various other approaches including:
 * Having the input be the negatives of the off-diagonal data of a Cartan matrix which are thus all natural numbers so we completely avoid `int.to_nat` and hard-code the diagonal 2.
 * Avoiding the `i = j` by using `set.image` on the off-diagonal of `B × B` instead of `set.range` when I take the union of all the relations

and I ended up convincing myself that what I have was the best compromise but dissenting opinions welcome!

Reply:
I think you could add `+1` after the `to_nat`, instead of inside. But it doesn't matter too much.

Reply:
> I think you could add `+1` after the `to_nat`, instead of inside. But it doesn't matter too much.

Yes indeed, I should have listed this too.